pixeltable/func/function.py

```diff
@@ -64,79 +73,189 @@ def display_name(self) -> str:
             return self.self_path[len(ptf_prefix):]
         return self.self_path
 
+    @property
+    def is_polymorphic(self) -> bool:
+        return len(self.signatures) > 1
+
+    @property
+    def signature(self) -> Signature:
+        assert not self.is_polymorphic
+        return self.signatures[0]
+
     @property
     def arity(self) -> int:
-        return len(self.signature.parameters)
+        assert not self.is_polymorphic
+        return len(self.signatures[0].parameters)
+
+    def _overload_resolution(self, signature_idx: int) -> Function:
+        """
+        Return a new  `Function` that retains just the single signature at index `signature_idx`, and is otherwise
+        identical to this `Function`.
+        """
+        assert 0 <= signature_idx < len(self.signatures)
+
+        # We cache the resolutions in self.__resolutions. This ensures that each resolution is represented globally
+        # by a single Python object. We do this dynamically rather than pre-constructing them in order to avoid
+        # circular complexity in the `Function` initialization logic.
+
+        # Note that it is necessary to create a resolution even if this Function has only one signature, because
+        # there is no guarantee we have seen all the Function's @overload declarations at the time this is called.
+
+        while len(self.__resolutions) <= signature_idx:
```

Comment:
Is there a specific reason why this goes through all prior idxs? Or could you just make resolutions a dict[idx, Function]?